### PR TITLE
Add service for stock endpoints

### PR DESCRIPTION
**Pull request by saschaheinl-fft**
Currently our SDK does not provide a service for the `api/stocks` endpoint. This PR adds a service with methods for 

- CRUD stocks (GET, POST, PUT, DELETE)
- GET stock summaries
- GET stock distribution

### CODE REVIEW

**Comment by arnoerpenbeck** [comment]
The new stock module must be exported in `fft-api/index.ts`

File: src/fft-api/stock/fftStockService.ts [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+        if (tenantArticleIds.length > 1) {
+          let query = `${tenantArticleIds[0]}&tenantArticleIds=`;
+          tenantArticleIds = tenantArticleIds.slice(1, 499);
+          query += tenantArticleIds.join('&tenantArticleIds=');
+
+          queryParams['tenantArticleIds'] = query;
+        }
+      }
+
```
Comment (nwilhelmFFT) [inline src/fft-api/stock/fftStockService.ts:68]: Maybe we could introduce a shared function for building the query so we can reuse it in multiple places.

File: src/fft-api/stock/fftStockService.ts [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        if (tenantArticleIds.length > 1) {
+          let query = `${tenantArticleIds[0]}&tenantArticleId=`;
+          tenantArticleIds = tenantArticleIds.slice(1, 499);
+          query += tenantArticleIds.join('&tenantArticleId=');
+
+          queryParams['tenantArticleId'] = query;
+        }
+      }
+
+      return this.apiClient.get<StockSummaries>(this.path, queryParams);
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:249]: use correct path for stocks summaries

File: src/fft-api/stock/fftStockService.ts [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        if (channelRefs.length > 1) {
+          let query = `${channelRefs[0]}&channelRefs=`;
+          channelRefs = channelRefs.slice(1);
+          query += channelRefs.join('&channelRefs=');
+
+          queryParams['channelRefs'] = query;
+        }
+      }
+
+      return this.apiClient.get<StockDistribution>(this.path);
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:341]: use correct path for stocks distribution, include query params

File: src/fft-api/stock/fftStockService.ts [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        if (tenantArticleIds.length == 1) {
+          queryParams['tenantArticleIds'] = tenantArticleIds[0];
+        }
+
+        if (tenantArticleIds.length > 1) {
+          let query = `${tenantArticleIds[0]}&tenantArticleIds=`;
+          tenantArticleIds = tenantArticleIds.slice(1, 499);
+          query += tenantArticleIds.join('&tenantArticleIds=');
+
+          queryParams['tenantArticleIds'] = query;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:65]: API param is called `tenantArticleId`

File: src/fft-api/stock/fftStockService.ts [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+        }
+      }
+
+      if (locationRefs) {
+        locationRefs = locationRefs.slice(0, 499);
+
+        if (locationRefs.length == 1) {
+          queryParams['locationRefs'] = locationRefs[0];
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:73]: API param is called `locationRef`

File: src/fft-api/stock/fftStockService.ts [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+        if (serviceTypes.length == 1) {
+          queryParams['facilityServiceTypes'] = serviceTypes[0];
+        }
+
+        if (serviceTypes.length > 1) {
+          let query = `${serviceTypes[0]}&facilityServiceType=`;
+          serviceTypes = serviceTypes.slice(1);
+          query += serviceTypes.join('&facilityServiceType=');
+
+          queryParams['facilityServiceType'] = query;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:195]: `facilityServiceTypes`

File: src/fft-api/stock/fftStockService.ts [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+      if (allowStale) {
+        queryParams['allowStale'] = String(allowStale);
+      }
+
+      if (tenantArticleIds) {
+        tenantArticleIds = tenantArticleIds.slice(0, 499);
+
+        if (tenantArticleIds.length == 1) {
+          queryParams['tenantArticleId'] = tenantArticleIds[0];
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:237]: `tenantArticleIds`

**Review by arnoerpenbeck — CHANGES_REQUESTED** [review]
need to fix request path and param names

File: src/fft-api/stock/fftStockService.ts [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+        }
+
+        if (tenantArticleIds.length > 1) {
+          let query = `${tenantArticleIds[0]}&tenantArticleIds=`;
```
Comment (nwilhelmFFT) [inline src/fft-api/stock/fftStockService.ts:62]: Why do you not use buildQueryFromArray here?

File: src/fft-api/stock/fftStockService.ts [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+  public async getStockSummaries(
+    size?: number,
+    startAfterId?: string,
+    serviceTypes?: FacilityServiceType[],
+    facilityStatuses?: FacilityStatus[],
+    facilityRefs?: string[],
+    allowStale?: boolean,
+    tenantArticleIds?: string[]
+  ): Promise<StockSummaries> {
```
Comment (nwilhelmFFT) [inline src/fft-api/stock/fftStockService.ts:174]: Should we also include channelRefs ?

File: src/fft-api/stock/fftStockService.ts [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+      if (tenantArticleIds) {
+        tenantArticleIds = tenantArticleIds.slice(0, 499);
+
+        if (tenantArticleIds.length == 1) {
+          queryParams['tenantArticleId'] = tenantArticleIds[0];
+        }
+
+        if (tenantArticleIds.length > 1) {
+          queryParams['tenantArticleIds'] = buildQueryFromArray('tenantArticleIds', tenantArticleIds, 500);
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:62]: `tenantArticleId`

Comment (saschaheinl-fft) [reply src/fft-api/stock/fftStockService.ts]: Done

File: src/fft-api/stock/fftStockService.ts [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+  ): Promise<StockPaginatedResult> {
+    try {
+      const queryParams: Record<string, string> = {};
+
+      if (facilityRef) {
+        queryParams['facilityRef'] = facilityRef;
+      }
+
+      if (tenantArticleIds) {
+        tenantArticleIds = tenantArticleIds.slice(0, 499);
```
Comment (nwilhelmFFT) [inline src/fft-api/stock/fftStockService.ts:55]: The slicing seems to already be done in `buildQueryFromArray`

File: src/fft-api/stock/fftStockService.ts [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+      if (locationRefs) {
+        locationRefs = locationRefs.slice(0, 499);
+
+        if (locationRefs.length == 1) {
+          queryParams['locationRef'] = locationRefs[0];
+        }
+
+        if (locationRefs.length > 1) {
+          let query = `${locationRefs[0]}&locationRef=`;
```
Comment (nwilhelmFFT) [inline src/fft-api/stock/fftStockService.ts:74]: You could also use `buildQueryFromArray` here

File: src/common/httpClient/httpClient.ts [hunk 13]
```diff
@@ -22,6 +22,9 @@ export class HttpClient implements BasicHttpClient {
     }
 
     if (config.params) {
+      if (Array.isArray(config.params)) {
+        request.query(config.params.join('&'));
+      }
       request.query(config.params);
```
Comment (arnoerpenbeck) [inline src/common/httpClient/httpClient.ts:28]: fehlt hier nicht ein `else`?

Comment (saschaheinl-fft) [reply src/common/httpClient/httpClient.ts:28]: Nein, das `if` muss komplett weg. Pushe eben, danke für den Hinweis

File: src/fft-api/stock/fftStockService.ts [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+      if (size && size <= 100) {
+        queryParams['size'] = size.toString();
+      }
+
+      if (startAfterId) {
+        queryParams['startAfterId'] = startAfterId;
+      }
+
+      if (serviceTypes && serviceTypes.length > 0) {
+        queryParams['serviceTypes'] = serviceTypes;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:166]: `facilityServiceTypes`

File: src/fft-api/stock/fftStockService.ts [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+    tenantArticleIds?: string[],
+    locationRefs?: string[],
+    size?: number,
+    startAfterId?: string
+  ): Promise<StockPaginatedResult> {
+    try {
+      const queryParams: QueryParams = {};
+
+      if (facilityRef) {
+        queryParams['facilityRefs'] = facilityRef;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:50]: `facilityRef`

File: src/fft-api/stock/fftStockService.ts [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+      }
+
+      if (tenantArticleIds) {
+        tenantArticleIds = tenantArticleIds.slice(0, 499);
+        queryParams['tenantArticleIds'] = tenantArticleIds;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:55]: `tenantArticleId`

File: src/fft-api/stock/fftStockService.ts [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+      }
+
+      if (locationRefs) {
+        locationRefs = locationRefs.slice(0, 499);
+        queryParams['locationRefs'] = locationRefs;
```
Comment (arnoerpenbeck) [inline src/fft-api/stock/fftStockService.ts:60]: `locationRef`

**Review by arnoerpenbeck — CHANGES_REQUESTED** [review]
some parameter names are still wrong

**Review by arnoerpenbeck — APPROVED** [review]

**Review by nwilhelmFFT — APPROVED** [review]

**Pull request merged**